File: src/org/lorenzoleonardini/badger/physics/PhysicsEngine.java
package org.lorenzoleonardini.badger.physics;

import java.util.ArrayList;
import java.util.List;

import org.lorenzoleonardini.badger.Engine;

/**
 * @author Lorenzo Leonardini
 */
public class PhysicsEngine
{
	private Vector2D gravity = new Vector2D(0, 5);

	private List<PhysicsObject> toRemove = new ArrayList<PhysicsObject>();

	private Engine engine;

	public World world;

	public PhysicsEngine(Engine engine)
	{
		this.engine = engine;
		world = new World();
	}

	public void update()
	{
		List<PhysicsObject> objs = new ArrayList<PhysicsObject>(ObjectManager.getObjects());

		for (PhysicsObject o : objs)
		{
			if (o.isAffectedByGravity() && !o.isUnmovable())
				o.applyForce(gravity);

			o.update(engine.getDelta());

			if (o.killWhenOutOfBounds())
				if (o.position.y - o.objectRender.h >= engine.HEIGHT)
					toRemove.add(o);

			if (!toRemove.contains(o))
				if (o.toBeKilled())
					toRemove.add(o);
		}

		ObjectManager.getObjects().removeAll(toRemove);

		for (PhysicsObject o : objs)
		{
			o.collide(world.getNeightbours(o));
		}
	}
}

File: Badger-core/src/physics/physics_engine.cpp
#include "physics_engine.h"

namespace badger { namespace physics {

	std::vector<Entity*> PhysicsEngine::ENTITY_LIST;

	void PhysicsEngine::update(double delta)
	{
		if (ENTITY_LIST.size() == 0)
			return;
		
		for (int i = 0; i < ENTITY_LIST.size(); i++)
		{
			Entity *e = ENTITY_LIST[i];

			if (e->isAffectedByGravity() && !e->isUnmovable())
				e->applyForce(m_Gravity);

			e->update(delta);

			if (e->toBeKilled() || (e->killWhenOutOfBounds() && e->m_Renderable->getPosition().y < -260 - e->m_Renderable->getSize().y))
			{
				//TODO : DELETE
				//delete ENTITY_LIST[i];
				delete e;
				ENTITY_LIST.erase(ENTITY_LIST.begin() + i);
			}
		}

		CircleHitbox::circleCircleCollision();

		//TODO: collision
		//for (int i = 0; i < ENTITY_LIST.size(); i++)
		//{
		//	Entity *e = ENTITY_LIST[i];
		//}
	}

} }